app/models.py

```diff
@@ -948,13 +1327,142 @@ def query_records_accessible_by(
             'delete']`.
         options: list of `sqlalchemy.orm.MapperOption`s
-           Options that will be passed to `options()` in the loader query.
+            Options that will be passed to `options()` in the loader query.
+        columns: list of sqlalchemy.Column, optional, default None
+            The columns to retrieve from the target table. If None, queries
+            the mapped class directly and returns mapped instances.
 
         Returns
         -------
         query: sqlalchemy.Query
             The query for the specified records.
+        """
+
+        if not isinstance(user_or_token, (User, Token)):
+            raise ValueError(
+                "user_or_token must be an instance of User or Token, "
+                f"got {user_or_token.__class__.__name__}."
+            )
+
+        logic = getattr(cls, mode)
+        return logic.query_accessible_rows(cls, user_or_token, columns=columns).options(
+            options
+        )
+
+    @classmethod
+    def get(
+        cls,
+        id_or_list,
+        user_or_token,
+        mode="read",
+        raise_if_none=False,
+        options=[],
+    ):
+        """Return a database record if it is accessible to the specified User or
+        Token. If no record exists, return None. If the record exists but is
+        inaccessible, raise an `AccessError`.
+        Parameters
+        ----------
+        id_or_list: int, str, iterable of int, iterable of str
+            The primary key(s) of the record(s) to query for.
+        user_or_token: `baselayer.app.models.User` or `baselayer.app.models.Token`
+            The User or Token to check.
+        mode: string
+            Type of access to check. Valid choices are `['create', 'read', 'update',
+            'delete']`.
+        options: list of `sqlalchemy.orm.MapperOption`s
+           Options that will be passed to `options()` in the loader query.
+
+        Returns
+        -------
+        record: `baselayer.app.models.Base` or list of `baselayer.app.models.Base`
+            The requested record(s). Has the same shape as `id_or_list`.
+        """
+
+        original_shape = np.asarray(id_or_list).shape
+        standardized = np.atleast_1d(id_or_list)
+        result = []
+
+        with Session(user_or_token) as session:
+            # TODO: vectorize this
+            for pk in standardized:
+                if options:
+                    stmt = sa.select(cls).options(options).where(cls.id == pk.item())
+                else:
+                    stmt = sa.select(cls).where(cls.id == pk.item())
+                instance = session.scalars(stmt).first()
+                if raise_if_none:
+                    if instance is None or not instance.is_accessible_by(
+                        user_or_token, mode=mode
+                    ):
+                        raise AccessError(f"Cannot find {cls.__name__} with id: {pk}")
```

Comment:
Why do we raise here instead of just returning None?

Reply:
I've changed the access pattern here. It makes sure that if the record doesn't exist OR if it inaccessible the results are exactly the same.